File: apps/web/src/types/novel.ts
export type NovelStatus = "draft" | "in_progress" | "completed" | "archived";

export type { APIResponse } from "@/types/user";

export type ChapterStatus = "draft" | "in_progress" | "completed" | "failed";

export type GenerationStatus =
  | "waiting"
  | "generating"
  | "completed"
  | "failed"
  | "skipped";

export interface Novel {
  id: string;
  user_id: string;
  title: string;
  premise: string | null;
  genre: string | null;
  synopsis: string | null;
  cover_path: string | null;
  blurb: string | null;
  word_count_target: number | null;
  status: NovelStatus;
  created_at: string;
  updated_at: string;
  chapters?: Chapter[];
  characters?: Character[];
  settings?: Setting[];
}

export interface Chapter {
  id: string;
  novel_id: string;
  title: string;
  outline: string | null;
  content: string | null;
  word_count: number;
  order_index: number;
  status: ChapterStatus;
  generation_status: GenerationStatus;
  created_at: string;
  updated_at: string;
  note?: ChapterNote;
}

export interface Character {
  id: string;
  novel_id: string;
  name: string;
  description: string | null;
  image_path: string | null;
  created_at: string;
  updated_at: string;
}

export interface Setting {
  id: string;
  novel_id: string;
  name: string;
  description: string | null;
  image_path: string | null;
  created_at: string;
  updated_at: string;
}

export interface ChapterNote {
  id: string;
  chapter_id: string;
  content: string;
  created_at: string;
  updated_at: string;
}

export interface CreateNovelInput {
  title: string;
  premise?: string;
  genre?: string;
  synopsis?: string;
}

export interface UpdateNovelInput {
  title?: string;
  premise?: string;
  genre?: string;
  synopsis?: string;
  blurb?: string;
}

export interface CreateChapterInput {
  title: string;
  outline?: string;
  order_index?: number;
}

export interface UpdateChapterInput {
  title?: string;
  outline?: string;
}

export interface UpdateChapterContentInput {
  content: string;
}

export interface ReorderChaptersInput {
  chapter_ids: string[];
}

export interface CreateCharacterInput {
  name: string;
  description?: string;
}

export interface UpdateCharacterInput {
  name?: string;
  description?: string;
}

export interface CreateSettingInput {
  name: string;
  description?: string;
}

export interface UpdateSettingInput {
  name?: string;
  description?: string;
}

export interface UpdateNoteInput {
  content: string;
}

File: api/internal/models/novel.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// NovelStatus enum: draft, in_progress, completed, archived
type NovelStatus string

const (
	NovelStatusDraft     NovelStatus = "draft"
	NovelStatusInProgess NovelStatus = "in_progress"
	NovelStatusCompleted NovelStatus = "completed"
	NovelStatusArchived  NovelStatus = "archived"
)

// Novel adalah project novel milik user, root entity untuk semua data novel.
// Status di-update otomatis: draft → in_progress (saat mulai menulis)
// → completed (saat semua bab selesai).
type Novel struct {
	ID             uuid.UUID   `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	UserID         uuid.UUID   `gorm:"type:uuid;not null;index" json:"user_id"`
	Title          string      `gorm:"type:varchar(200);not null" json:"title"`
	Premise        *string     `gorm:"type:varchar(500)" json:"premise"`
	Genre          *string     `gorm:"type:varchar(50)" json:"genre"`
	Synopsis       *string     `gorm:"type:text" json:"synopsis"`
	CoverPath      *string     `gorm:"type:varchar(500)" json:"cover_path"`
	Blurb          *string     `gorm:"type:text" json:"blurb"`
	WordCountTarget *int       `gorm:"type:integer" json:"word_count_target"`
	Status         NovelStatus `gorm:"type:novel_status;not null;default:'draft'" json:"status"`
	CreatedAt      time.Time   `gorm:"not null;default:now()" json:"created_at"`
	UpdatedAt      time.Time   `gorm:"not null;default:now()" json:"updated_at"`

	// Relations
	Chapters    []Chapter    `gorm:"foreignKey:NovelID" json:"chapters,omitempty"`
	Characters  []Character  `gorm:"foreignKey:NovelID" json:"characters,omitempty"`
	Settings    []Setting    `gorm:"foreignKey:NovelID" json:"settings,omitempty"`
	FileAssets  []FileAsset  `gorm:"foreignKey:NovelID" json:"-"`
	Exports     []NovelExport `gorm:"foreignKey:NovelID" json:"-"`
}

func (Novel) TableName() string {
	return "novels"
}
